### PR TITLE
Features/actor serialization

### PR DESCRIPTION
**Pull request by sfc-gh-mpilman**

### CODE REVIEW

File: fdbclient/ActorLineageProfiler.cpp [hunk 1]
```diff
@@ -0,0 +1,183 @@
+/*
+ * ActorLineageProfiler.h
```
Comment (sfc-gh-ljoswiak) [inline fdbclient/ActorLineageProfiler.cpp]:
```suggestion
 * ActorLineageProfiler.cpp
```

File: fdbclient/ActorLineageProfiler.cpp [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2013-20201 Apple Inc. and the FoundationDB project authors
```
Comment (sfc-gh-ljoswiak) [inline fdbclient/ActorLineageProfiler.cpp]:
```suggestion
 * Copyright 2013-2021 Apple Inc. and the FoundationDB project authors
```

File: fdbclient/ActorLineageProfiler.h [hunk 3]
```diff
@@ -0,0 +1,80 @@
+/*
+ * ActorLineageProfiler.h
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2013-20201 Apple Inc. and the FoundationDB project authors
```
Comment (sfc-gh-ljoswiak) [inline fdbclient/ActorLineageProfiler.h]:
```suggestion
 * Copyright 2013-2021 Apple Inc. and the FoundationDB project authors
```

File: fdbclient/ActorLineageProfiler.cpp [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+		}
+	}
+
+	void pack(int64_t val) {
+		if (val >= 0) {
+			this->pack(uint64_t(val));
+		} else if (val >= std::numeric_limits<uint8_t>::min()) {
+			pack_int8(int8_t(val));
+		} else if (val >= std::numeric_limits<uint16_t>::min()) {
+			pack_int8(int16_t(val));
```
Comment (sfc-gh-ljoswiak) [inline fdbclient/ActorLineageProfiler.cpp]:
```suggestion
			pack_int16(int16_t(val));
```

File: fdbclient/ActorLineageProfiler.cpp [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+		} else if (val >= std::numeric_limits<uint32_t>::min()) {
+			pack_int8(int32_t(val));
```
Comment (sfc-gh-ljoswiak) [inline fdbclient/ActorLineageProfiler.cpp]:
```suggestion
			pack_int32(int32_t(val));
```

File: fdbclient/ActorLineageProfiler.cpp [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+		} else if (val >= std::numeric_limits<uint64_t>::min()) {
+			pack_int8(int64_t(val));
```
Comment (sfc-gh-ljoswiak) [inline fdbclient/ActorLineageProfiler.cpp]:
```suggestion
			pack_int64(int64_t(val));
```

File: fdbclient/ActorLineageProfiler.cpp [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+			pack_uint64(val);
+		}
+	}
+
+	void pack(int64_t val) {
+		if (val >= 0) {
+			this->pack(uint64_t(val));
+		} else if (val >= std::numeric_limits<uint8_t>::min()) {
+			pack_int8(int8_t(val));
+		} else if (val >= std::numeric_limits<uint16_t>::min()) {
```
Comment (sfc-gh-ljoswiak) [inline fdbclient/ActorLineageProfiler.cpp]:
```suggestion
		} else if (val >= std::numeric_limits<int16_t>::min()) {
```

File: fdbclient/ActorLineageProfiler.cpp [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+			pack_int8(int16_t(val));
+		} else if (val >= std::numeric_limits<uint32_t>::min()) {
```
Comment (sfc-gh-ljoswiak) [inline fdbclient/ActorLineageProfiler.cpp]:
```suggestion
		} else if (val >= std::numeric_limits<int32_t>::min()) {
```

File: fdbclient/ActorLineageProfiler.cpp [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+			pack_int8(int32_t(val));
+		} else if (val >= std::numeric_limits<uint64_t>::min()) {
```
Comment (sfc-gh-ljoswiak) [inline fdbclient/ActorLineageProfiler.cpp]:
```suggestion
		} else if (val >= std::numeric_limits<int64_t>::min()) {
```

File: fdbclient/ActorLineageProfiler.h [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+	using Getter = std::function<std::vector<Reference<ActorLineage>>()>;
+
+private:
+	std::vector<IALPCollectorBase*> collectors;
+	std::map<WaitState, Getter> getSamples;
+	SampleCollectorT() {}
+
+public:
+	void addCollector(IALPCollectorBase* collector) { collectors.push_back(collector); }
+	std::map<std::string_view, std::any> collect(ActorLineage* lineage);
```
Comment (sfc-gh-ljoswiak) [inline fdbclient/ActorLineageProfiler.h]: Can this function be private? Will anyone besides the other `collect` function here want to call this `collect`?

**Review by sfc-gh-ljoswiak — CHANGES_REQUESTED** [review]

File: fdbclient/ActorLineageProfiler.cpp [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+				map.emplace(std::type_index(typeid(Head)), any_visitor<Head>);
+				populate_visitor_map<Tail...>::populate(map);
+			}
+		};
+		template <>
+		struct populate_visitor_map<> {
+			static void populate(VisitorMap&) {}
+		};
+
+		visitor_t() { populate_visitor_map<int64_t, uint64_t, bool, float, double>::populate(visitorMap); }
```
Comment (sfc-gh-ljoswiak) [inline fdbclient/ActorLineageProfiler.cpp]: Missing string, vector, and map types.

File: fdbclient/ActorLineageProfiler.h (hunk 10 above)
Comment (sfc-gh-mpilman) [reply fdbclient/ActorLineageProfiler.h]: yes, I can make this private.

**Review by sfc-gh-ljoswiak — APPROVED** [review]

**Pull request merged**